libs/datamodel/core/src/transform/ast_to_dml/validate/referential_actions.rs

```diff
@@ -1,63 +1,148 @@
+                continue;
+            }
 
-impl<'a> VisitedRelation<'a> {
-    /// Create a new root node, starting a new relation path.
-    fn root(model_name: &'a str, field_name: &'a str) -> Self {
-        Self {
-            previous: None,
-            model_name,
-            field_name,
+            // Traversing all other relations from the next model.
+            related_model
+                .relation_fields()
+                .filter(|f| f.is_singular())
+                .for_each(|related_field| {
+                    next_relations.push((
+                        related_model,
+                        related_field,
+                        Rc::new(visited_relations.link_next(related_model.name(), related_field.name())),
+                    ));
+                });
         }
     }
 
-    /// Links a relation to the current path.
-    fn link_next(self: &Rc<Self>, model_name: &'a str, field_name: &'a str) -> Self {
-        Self {
-            previous: Some(self.clone()),
-            model_name,
-            field_name,
+    // Gather all paths from a relation field to all the models seen from that
+    // field.
+    let mut seen: HashMap<&str, HashSet<&str>> = HashMap::new();
+    for path in paths.iter() {
+        let mut iter = path.iter();
+
+        let seen_models = match iter.next() {
+            Some((_, Some(field_name))) => seen.entry(field_name).or_default(),
+            _ => continue,
+        };
+
+        for (model_name, _) in iter {
+            seen_models.insert(model_name);
         }
     }
-}
 
-impl<'a> fmt::Display for VisitedRelation<'a> {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        let mut traversed_models = vec![format!("{}.{}", self.model_name, self.field_name)];
-        let mut this = self;
+    // Compact the model nameas that can be reached from the given relation
+    // field, but also from any other relation field in the same model.
+    let mut reachable: BTreeSet<&str> = BTreeSet::new();
 
-        while let Some(next) = this.previous.as_ref() {
-            traversed_models.push(format!("{}.{}", next.model_name, next.field_name));
-            this = next;
+    if let Some(from_parent) = seen.remove(parent_field.name().as_str()) {
+        for (_, from_other) in seen.into_iter() {
+            reachable.extend(from_parent.intersection(&from_other));
         }
+    }
 
-        traversed_models.reverse();
+    let models = reachable.iter().map(|model| format!("`{}`", model)).join(", ");
 
-        write!(f, "{}", traversed_models.join(" → "))
+    match reachable.len() {
+        0 => (),
+        1 => {
+            let msg = format!(
+                "When any of the records in model {} is updated or deleted, the referential actions on the relations cascade to model `{}` through multiple paths. Please break one of these paths by setting the `onUpdate` and `onDelete` to `NoAction`.", models, parent_model.name());
```

Comment:
I am wondering, would it be possible to say if the path is made of onDeletes, onUpdates or both here?